File: go/test/today_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/useless-facts-sdk/go"
	"github.com/voxgig-sdk/useless-facts-sdk/go/core"

	vs "github.com/voxgig-sdk/useless-facts-sdk/go/utility/struct"
)

func TestTodayEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Today(nil)
		if ent == nil {
			t.Fatal("expected non-nil TodayEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := todayBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "today." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set USELESSFACTS_TEST_TODAY_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		todayRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.today", setup.data)))
		var todayRef01Data map[string]any
		if len(todayRef01DataRaw) > 0 {
			todayRef01Data = core.ToMapAny(todayRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = todayRef01Data

		// LOAD
		todayRef01Ent := client.Today(nil)
		todayRef01MatchDt0 := map[string]any{
			"id": todayRef01Data["id"],
		}
		todayRef01DataDt0Loaded, err := todayRef01Ent.Load(todayRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		todayRef01DataDt0LoadResult := core.ToMapAny(todayRef01DataDt0Loaded)
		if todayRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if todayRef01DataDt0LoadResult["id"] != todayRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func todayBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "today", "TodayTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read today test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse today test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"today01", "today02", "today03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("USELESSFACTS_TEST_TODAY_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"USELESSFACTS_TEST_TODAY_ENTID": idmap,
		"USELESSFACTS_TEST_LIVE":      "FALSE",
		"USELESSFACTS_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["USELESSFACTS_TEST_TODAY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["USELESSFACTS_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewUselessFactsSDK(core.ToMapAny(mergedOpts))
	}

	live := env["USELESSFACTS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["USELESSFACTS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/today_entity_test.rb
# Today entity test

require "minitest/autorun"
require "json"
require_relative "../UselessFacts_sdk"
require_relative "runner"

class TodayEntityTest < Minitest::Test
  def test_create_instance
    testsdk = UselessFactsSDK.test(nil, nil)
    ent = testsdk.Today(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = today_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "today." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set USELESS_FACTS_TEST_TODAY_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    today_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.today")))
    today_ref01_data = nil
    if today_ref01_data_raw.length > 0
      today_ref01_data = Helpers.to_map(today_ref01_data_raw[0][1])
    end

    # LOAD
    today_ref01_ent = client.Today(nil)
    today_ref01_match_dt0 = {
      "id" => today_ref01_data["id"],
    }
    today_ref01_data_dt0_loaded = today_ref01_ent.load(today_ref01_match_dt0, nil)
    today_ref01_data_dt0_load_result = Helpers.to_map(today_ref01_data_dt0_loaded.respond_to?(:data_get) ? today_ref01_data_dt0_loaded.data_get : today_ref01_data_dt0_loaded)
    assert !today_ref01_data_dt0_load_result.nil?
    assert_equal today_ref01_data_dt0_load_result["id"], today_ref01_data["id"]

  end
end

def today_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "today", "TodayTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = UselessFactsSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["today01", "today02", "today03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["USELESS_FACTS_TEST_TODAY_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "USELESS_FACTS_TEST_TODAY_ENTID" => idmap,
    "USELESS_FACTS_TEST_LIVE" => "FALSE",
    "USELESS_FACTS_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["USELESS_FACTS_TEST_TODAY_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["USELESS_FACTS_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = UselessFactsSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["USELESS_FACTS_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["USELESS_FACTS_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
